kaggle_environments/envs/halite/helpers.py

```diff
@@ -0,0 +1,781 @@
+from copy import deepcopy
+from enum import Enum, auto
+from typing import *
+
+# region Helper Classes and Methods
+Point = NewType('Point', Tuple[int, int])
+
+
+def translate_point(point: Point, offset: Point) -> Point:
+    """Returns (point.x + offset.x, point.y + offset.y)"""
+    (x1, y1) = point
+    (x2, y2) = offset
+    return x1 + x2, y1 + y2
+
+
+def wrap_point(point: Point, size: int) -> Point:
+    """
+    Returns (point.x % size, point.y % size)
+    If the point is not on the board of width and height size it will be wrapped around to fit on the board
+    """
+    (x, y) = point
+    return x % size, y % size
+
+
+def position_to_index(point: Point, size: int) -> int:
+    """
+    Converts a 2d position in the form (x, y) to an index in the observation.halite list.
+    See index_to_position for the inverse.
+    """
+    x, y = point
+    return (size - y - 1) * size + x
+
+
+def index_to_position(index: int, size: int) -> Point:
+    """
+    Converts an index in the observation.halite list to a 2d position in the form (x, y).
+    See position_to_index for the inverse.
+    """
+    y, x = divmod(index, size)
+    return x, (size - y - 1)
+
+
+TElement = TypeVar('TElement')
+THash = TypeVar('TComparable')
+
+
+def group_by(elements: Iterable[TElement], selector: Callable[[TElement], THash]) -> Dict[THash, List[TElement]]:
+    results = {}
+    for element in elements:
+        key = selector(element)
+        if key not in results:
+            results[key] = []
+        results[key].append(element)
+    return results
+
+
+TKey = TypeVar('TKey')
+TValue = TypeVar('TValue')
+
+
+class ReadOnlyDict(Generic[TKey, TValue]):
```

Comment:
Everything from here to the board class is mainly just data binding models. Reviewers can mostly skim these, but there are some small bits of logic like `ShipAction.to_point` and `Player.next_actions`